azurerm/resource_arm_azure_firewall_network_rule_collection.go

```diff
@@ -0,0 +1,355 @@
+	name := d.Get("name").(string)
+	action := network.AzureFirewallRCActionType(d.Get("action").(string))
+	rules := expandArmAzureFirewallNetworkRules(d)
+	properties := network.AzureFirewallNetworkRuleCollectionPropertiesFormat{
+		Action: &network.AzureFirewallRCAction{
+			Type: action,
+		},
+		Priority: utils.Int32(int32(d.Get("priority").(int))),
+		Rules:    &rules,
+	}
+	col := network.AzureFirewallNetworkRuleCollection{
+		Name: &name,
+		AzureFirewallNetworkRuleCollectionPropertiesFormat: &properties,
+	}
+	return col
+}
+
+func findArmAzureFirewallNetworkRuleCollectionByName(firewall *network.AzureFirewall, name string) (*network.AzureFirewallNetworkRuleCollection, bool) {
+	for _, collection := range *firewall.AzureFirewallPropertiesFormat.NetworkRuleCollections {
+		if collection.Name != nil && *collection.Name == name {
+			return &collection, true
+		}
+	}
+	return nil, false
+}
+
+func removeArmAzureFirewallNetworkRuleCollectionByName(firewall *network.AzureFirewall, name string) *[]network.AzureFirewallNetworkRuleCollection {
+	var index int
+	collections := *firewall.AzureFirewallPropertiesFormat.NetworkRuleCollections
+	for i, collection := range collections {
+		if collection.Name != nil && *collection.Name == name {
+			index = i
+			continue
+		}
+	}
+	collections = append(collections[:index], collections[index+1:]...)
+	return &collections
+}
+
+func expandArmAzureFirewallNetworkRules(d *schema.ResourceData) []network.AzureFirewallNetworkRule {
+	nwRules := d.Get("rule").(*schema.Set).List()
+	rules := make([]network.AzureFirewallNetworkRule, 0)
+
+	for _, nwRule := range nwRules {
+		rule := nwRule.(map[string]interface{})
+
+		name := rule["name"].(string)
+		ruleToAdd := network.AzureFirewallNetworkRule{
+			Name: &name,
+		}
+		if description := rule["description"].(string); description != "" {
+			ruleToAdd.Description = &description
+		}
+		if r, ok := rule["source_addresses"].(*schema.Set); ok && r.Len() > 0 {
+			ruleToAdd.SourceAddresses = expandArmAzureFirewallSet(r)
+		}
+		if r, ok := rule["destination_addresses"].(*schema.Set); ok && r.Len() > 0 {
+			ruleToAdd.DestinationAddresses = expandArmAzureFirewallSet(r)
+		}
+		if r, ok := rule["destination_ports"].(*schema.Set); ok && r.Len() > 0 {
+			ruleToAdd.DestinationPorts = expandArmAzureFirewallSet(r)
+		}
+		if r, ok := rule["protocols"].(*schema.Set); ok && r.Len() > 0 {
+			var protocols []network.AzureFirewallNetworkRuleProtocol
+			for _, v := range r.List() {
+				s := network.AzureFirewallNetworkRuleProtocol(v.(string))
+				protocols = append(protocols, s)
+			}
+			ruleToAdd.Protocols = &protocols
+		}
+		rules = append(rules, ruleToAdd)
+	}
+
+	return rules
+}
+
+func flattenArmAzureFirewallNetworkRules(rules *[]network.AzureFirewallNetworkRule) []map[string]interface{} {
+	result := make([]map[string]interface{}, 0)
+
+	if rules != nil {
```

Comment:
_minor_ we could probably invert this to make it more readable, but it's not a blocker by any means :)